src/NServiceBus.Metrics/RawData/RingBuffer.cs

```diff
@@ -0,0 +1,106 @@
+﻿namespace NServiceBus.Metrics.RawData
+{
+    using System;
+    using System.Threading;
+
+    class RingBuffer
+    {
+        public const int Size = 4096;
+        const long SizeMask = Size - 1;
+        const long EpochMask = ~SizeMask;
+
+        long nextToWrite;
+        long nextToConsume;
+
+        Entry[] entries = new Entry[Size];
+
+        public struct Entry
+        {
+            public Entry(long ticks, long value)
+            {
+                Ticks = ticks;
+                Value = value;
+            }
+
+            public long Ticks;
+            public long Value;
+        }
+
+        public bool TryWrite(long value)
+        {
+            var readNextWrite = Volatile.Read(ref nextToWrite);
+            long index;
+
+            do
+            {
+                var consume = Volatile.Read(ref nextToConsume);
+
+                // if writers overlaps with not consumed writes, end
+                if (readNextWrite - consume >= Size)
+                {
+                    return false;
+                }
+                index = readNextWrite;
+
+                // try to swap nextToWrite
+                readNextWrite = Interlocked.CompareExchange(ref nextToWrite, index + 1, index);
+
+                // do until the swap not succeeded
+            } while (index != readNextWrite);
+            
+            // index is claimed, writing data
+            var i = index & SizeMask;
+            var ticks = DateTime.Now.Ticks;
+
+            entries[i].Value = value;
+            Volatile.Write(ref entries[i].Ticks, ticks);
```

Comment:
Oh I see, where this comment about `var ` comes from. Entries are `struct`s not `class`es, so in this C# version I cannot obtain a `ref` to struct item of an array.